cmd/kubelet/app/server.go

```diff
@@ -186,6 +186,120 @@ func Run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies) error {
 	return nil
 }
 
+// setKubeDepsClients initializes clients if not standalone mode and any of the clients are not provided
+func setKubeDepsClients(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, nodeName types.NodeName) error {
+	clientConfig, err := CreateAPIServerClientConfig(s)
+	if err == nil {
+		clientCertificateManager, err := makeKubeletClientCertificateManager(s, clientConfig, nodeName)
+		if err != nil {
+			return err
+		}
+
+		kubeClient, err := makeKubeDepsKubeClient(kubeDeps, clientConfig)
+		if err != nil {
+			startClientCertificateManager(kubeClient, clientCertificateManager)
+		}
+
+		setKubeDepsExternalKubeClient(kubeDeps, clientConfig)
+		setKubeDepsEventClient(s, kubeDeps, clientConfig)
+		setKubeDepsHeartbeatClient(s, kubeDeps, clientConfig)
+
+	} else {
+		switch {
+		case s.RequireKubeConfig:
+			return fmt.Errorf("invalid kubeconfig: %v", err)
+		case s.KubeConfig.Provided():
+			glog.Warningf("invalid kubeconfig: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func makeKubeletClientCertificateManager(
+	s *options.KubeletServer,
+	clientConfig *restclient.Config,
+	nodeName types.NodeName) (certificate.Manager, error) {
+
+	var clientCertificateManager certificate.Manager
+	var err error
+	if s.RotateCertificates && utilfeature.DefaultFeatureGate.Enabled(features.RotateKubeletClientCertificate) {
+		clientCertificateManager, err = kubeletcertificate.NewKubeletClientCertificateManager(
+			s.CertDirectory,
+			nodeName,
+			clientConfig.CertData,
+			clientConfig.KeyData,
+			clientConfig.CertFile,
+			clientConfig.KeyFile)
+		if err != nil {
+			return nil, err
+		}
+		// we set exitIfExpired to true because we use this client configuration to request new certs - if we are unable
+		// to request new certs, we will be unable to continue normal operation
+		if err := kubeletcertificate.UpdateTransport(wait.NeverStop, clientConfig, clientCertificateManager, true); err != nil {
+			return nil, err
+		}
+	}
+
+	return clientCertificateManager, nil
+}
+
+func makeKubeDepsKubeClient(kubeDeps *kubelet.Dependencies, clientConfig *restclient.Config) (clientset.Interface, error) {
```

Comment:
Try to avoid the assignment of clients in constructor functions.

Why not have:
- `newKubeClient`
- `newExternalKubeClient`
- `newEventClient`
- `newHeartbeatClient`

which each return a client or an error, and then assign the values to kubeDeps in `setKubeDepsClients`?

Reply:
ok, good idea!